### PR TITLE
add filter for intermediate resonance particles

### PR DESCRIPTION
**Pull request by jieunyoo**
Add this GenFilter. There is a similar LHEPtFilter, but it only filters outgoing particles. This filter does the filtering on intermediate particles. For example, for a boosted Higgs, we would like to filter on the pT of the Higgs (not the final state partons). The current LHEPtFilter does not do that. This doesn't affect any other PRs or external code. 

(An attempt to extend the current LHEPtFilter was made instead of adding this filter, but difficulties were encountered with the plugin and I hope to get this addition added soon so we can ask for a small production of the sample.)

I would like to backport this also to 10_6_X to that it can be used for a Run 2 UL thesis.

### CODE REVIEW

**Comment by jieunyoo** [comment]
@cmsbuild please test

**Comment by jieunyoo** [comment]
@menglu21 Thanks!

**Comment by jieunyoo** [comment]
This filter works simply like this:

LHEHiggsPtFilter = cms.EDFilter("LHEIntermediateParticlePtFilter",
  selectedPdgIds = cms.vint32(25),
  ptMin=cms.double(135.),
  ptMax=cms.double(1e10),
  src=cms.InputTag("externalLHEProducer")
  
  A small private sample was made. This shows the filtered pT of the Higgs.
higgsPT-filtered

**Comment by menglu21** [comment]
@cmsbuild please test

**Review by perrotta — COMMENTED** [review]
Since this is a new filter, please also add a fillDescriptions method to define and validate the parameters

File: GeneratorInterface/GenFilters/plugins/LHEIntermediateParticlePtFilter.cc [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+ Description: Filter to select events with pT in a given range.
+ (Based on LHEGenericFilter)
+
+     
+*/
+//
+
+// system include files
+#include <memory>
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (jieunyoo) [reply GeneratorInterface/GenFilters/plugins/LHEIntermediateParticlePtFilter.cc:17]: removed iostream; other filters have memory so kept it

[the remark this answers is not in the document]
Comment (jieunyoo) [reply GeneratorInterface/GenFilters/plugins/LHEIntermediateParticlePtFilter.cc:17]: From checking other filters, it seems they had an "implementation section" in the comments. I added that section. Hope that addresses the above "fillDescriptions method." (I didn't see any specific fillDescriptions example.) Thanks for the quick review!

**Comment by perrotta** [comment]
@jieunyoo 
- you did not remove `iostream` in fact...
- fillDescriptions is a method used to validate and initialize the configuration parameters, see https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp . If you don't know how to implement it, don't mind: it is a rather simple filter, we can even skip validation...
- At the end of your fixes, please squash all commits into one so that you can clean up the history of the file

**Comment by jieunyoo** [comment]
Hi @perrotta . I squashed the commits, and removed "iostream." For this commit then (since it's my first ever to CMSSW), I guess I will skip the fillDescriptions method that you mentioned above (you mentioned it would be ok to skip). Thanks for the reference though and I'll keep it in mind for later. Does everything look good for you? Do I need to do anything else for a backport to 10_6_X for UL studies? I am not sure of the timeline. Hope to get this to my MC contact for a UL MC production. Thanks!

**Comment by perrotta** [comment]
Please test

**Comment by perrotta** [comment]
+1

**Pull request merged**